keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1069,187 @@ the remaining requests in that queue start getting faster service.  In
+the virtual world for `additional latency`.
+Adjusting virtual time of a queue to do that is trivial.  The other thing
+to tweak is to ensure that the concurrency units will not get available
+for other requests for that time (because currently all actions are
+triggerred by starting or finishing some request).  We will maintain that
+possibility by wrapping the handler into another one that will be sleeping
+for `additional latence` after the request is processed.
+
+Note that given the estimation for duration of processing the requests is
+automatically corrected (both up and down), there is no need to change that
+in the initial version.
+
+#### Dispatching the request
+
+The hardest part of adding support for LIST requests is dispatching the
+requests.  Now in order to start processing a request, it has to accumulate
+`<width>` units of concurrency.
+
+The important requirement to recast now is fairness.  As soon a single
+request can consume more units of concurrency, the fairness is
+no longer about the number of requests from a given queue, but rather
+about number of consumed concurrency units.  This justifes the above
+definition of adjusting the cost of the request to now be equal to
+`<width> x <processing latency>` (instead of just `<processing latency>`).
+
+At the same time, we want to maximally utilize the available capacity.
+In other words, we want to minimize the time when some concurrency unit
+is not used, but there are requests at a given PL that could use it.
+
+In order to achieve the above goals, we are introducing the following
+modification to the current dispatching algorith:
+- as soon as we dispatch the first token (concurrency unit) to a queue
+  with LIST request at its front, all further tokens are also assigned
+  to this queue until the LIST request can be started.
+  Such approach allows us to not block too many tokens and avoid deadlocks
+  if we would be dispatching tokens to multiple LIST requests without having
+  enough of them for a given priority level.
+- however, to ensure fairness (especially over longer period of times)
+  the virtual time of the queue is now advanced by the new cost of the
+  request, i.e. by `<width> x <processing latency>` (as opposed to simply
+  `<processing latency>`).
+  This can be thought of as a negative budget that we accumulate (due to
+  consuming more token that we were supposed to consume) and ensure that
+  in the next round of dispatching the queues that didn't get as much
+  compute power will be chosen.
+
+However, now the queueing mechanism also requires adjustment.  So far,
+when putting a request into the queue, we were choosing the shortest queue.
+It worked, because it well proxied the total cost of processing all requests
+in that queue.
+After the above changes, the size of the queue is no longer correctly
+approximating the cost of processing the queued items.  Given that the
+total cost of processing a request is now `<width> x <processing latency>`,
```

Comment:
I think our priority is 1. fairness and 2. throughput. I think the queues are a fairness mechanism, if we were focused on throughput then maybe taking the magnitude / area of requests like this would be a good step in a greedy bin packing algorithm, but if we were doing that I don't think we would have queues at all.

So, I still think we should do what I said in my last comment.

I think if we want to have both fairness AND bin-packing, we should consider a two-step pipeline:
1. As today, incoming requests go into queues. Use the existing fairness algorithm to dequeue requests. But instead of executing them directly, put them into the "swap space". We always keep N requests in the swap space.
2. A bin-packing algorithm determines a way to arrange the requests in the swap space which minimizes execution time.
3. To maintain the fairness quality, there has to be a cap on the amount of time a request can stay in the swap space.

The bin-packing algorithm is basically playing tetris, except it can't rotate the pieces.

Again, I think in this first step we should do the absolute minimum to get the width concept in. We can experiment later and demonstrate incremental improvements.

Reply:
I guess there is some misunderstanding here - I think what you suggest here is waaaay more complicate that what I'm saying.

What I'm suggesting are very simple and local changes and they are a simple consequence of saying that the cost of the request (wrt fairnes) is now <width> x <duration> (as the request is occupying <width> seats instead of 1.

Reply:
I think I'm really hung up on the units not matching :)

I agree with what you're saying in the comments here, it's just not what I understood from reading the text earlier today-- maybe I needed more coffee.

Reply:
> I think I'm really hung up on the units not matching :)

See my comment above about the units.

> it's just not what I understood from reading the text earlier today-- maybe I needed more coffee.

This wasn't fully clear - I agree. I have adjusted the text [mostly lines 1225-1237] - hopefully it's cleaner now.

Reply:
Yes, it is clearer now, thanks.